### ISSUE TITLE
install and shell scripts should get service hostnames from a config file

### ISSUE THREAD

**Issue by ahelsing** [GENI-NSF/geni-portal#354]
Currently, the install and shell scripts use /bin/hostname to figure out the host name to use. When the web server has a CNAME, that won't work right. And we also want to still support distributing the services at some point.

So instead set up some configs / environment variables to set the right host names, for use in generating URLs / generating service URNs / etc.

 _Imported from trac ticket [#354](http://trac.gpolab.bbn.com/proto-ch/ticket/354),  created by ahelsing on 11-29-2012 at 15:55, last modified: 05-28-2013 at 14:47_

**Comment by tcmitchell** [comment]
For starters, remove some obsolete scripts:
 bin/create_pki_materials.sh
 bin/activate_project_lead.sh
 bin/revoke_project_lead.sh
 sign-pgeni3-proxy-req


_Trac comment by tmitchel (github user: tcmitchell) on 04-08-2013 at 11:55_

**Comment by tcmitchell** [comment]
This work is complete in development and will be part of a future release.

_Trac comment by tmitchel (github user: tcmitchell) on 05-08-2013 at 13:07_

**Comment by seledwards** [comment]
I sat with Chaos while she tried out the code on `tau-ceti.gpolab.bbn.com` otherwise known as `portal5.gpolab.bbn.com` and `ch5.gpolab.bbn.com`.

We encountered the following issues:
 * Her user URN reported in SliverStatus is urn:publicid:IDN+geni:gpo:portal+user+chaos but is listed on the Profile page as urn:publicid:IDN+ch5.gpolab.bbn.com+user+chaos
    * Perhaps this is an issue with the `omni_config` we are generating?
 * The comment field (in the `authorized_keys` file) on Chaos' key loaded on the node  says `chaos@portal.geni.net` and should be more like chaos@ch5.gpolab.bbn.com
    * Chaos suggests that the hostname is hardcoded in `proto-ch/portal/www/portal/generatesshkey.php`


There was also an issue with SliverStatus intermittently failing on pgeni3, but I doubt that's an issue with the portal.geni.net transition.



_Trac comment by sedwards (github user: seledwards) on 05-15-2013 at 16:23_

**Comment by seledwards** [comment]
Tried downloading and installing omni.  Was able to run `getversion` and some basic commands.  Tried creating a slice and got the following error:
```
$ ./omni.py -c ~/.gcf/omni_config.tauceti createslice setest
INFO:omni:Loading config file /home/geniuser/.gcf/omni_config.tauceti
INFO:omni:Using control framework portal
ERROR:omni.pgch:Failed to create new PG slice urn:publicid:IDN+ch5:chaostest+slice+setest: Register got slice URN with unknown authority ch5//chaostest (code 1)
ERROR:omni:Create Slice Failed for slice name setest.
WARNING:omni:   Try re-running with --debug for more information.
INFO:omni: ------------------------------------------------------------
INFO:omni: Completed createslice:

  Options as run:
		configfile: /home/geniuser/.gcf/omni_config.tauceti
		framework: portal
		project: chaostest

  Args: createslice setest

  Result Summary: Create Slice Failed for slice name setest.
 
INFO:omni: ============================================================
```

The meat of my `omni_config` generated by `omni-configure.py` looks like this:
```
[omni]
default_cf=portal
users=sedwards
default_project=chaostest

[sedwards]
urn=urn:publicid:IDN+ch5.gpolab.bbn.com+user+sedwards
keys=/home/geniuser/.ssh/tauceti_sedwards_id_rsa.pub

[portal]
type = pgch
ch = https://ch5.gpolab.bbn.com:8443/
sa = https://ch5.gpolab.bbn.com:8443/
cert = /home/geniuser/.ssl/sedwards_on_tauceti.pem
key = /home/geniuser/.ssl/sedwards_on_tauceti.pem
```

Perhaps this is due to omni determining the authority from the value of `ch` or `sa`?

_Trac comment by sedwards (github user: seledwards) on 05-15-2013 at 16:40_

**Comment by ahelsing** [comment]
framework_pgch gets the authority from the 'sa' hostname (see around line 112)

As a test, try supplying the full correct URN for the slice you want to create, and see if that works.

We may need to add an 'authority' attribute in the omni_config for talking to our CH.


_Trac comment by ahelsing on 05-15-2013 at 16:51_

**Comment by ahelsing** [comment]
Aaron: full URN test worked (though it gave me a warning of course):
```
jericho,[~/tmp/portal5],16:43(0)$ ~/src/gcf-2.2.1/src/omni.py -c ~/tmp/portal5/omni_config createslice urn:publicid:IDN+ch5.gpolab.bbn.com:chaostest+slice+ecgtest2
INFO:omni:Loading config file /home/chaos/tmp/portal5/omni_config
INFO:omni:Using control framework portal
WARNING:omni.pgch:CAREFUL: slice' authority (ch5.gpolab.bbn.com:chaostest) doesn't match current configured authority (ch5:chaostest)
INFO:omni.pgch:This may be OK though if you are using delegated slice credentials...
INFO:omni:Created slice with Name urn:publicid:IDN+ch5.gpolab.bbn.com:chaostest+slice+ecgtest2, URN urn:publicid:IDN+ch5.gpolab.bbn.com:chaostest+slice+ecgtest2, Expiration 2013-05-16 04:53:55
INFO:omni: ------------------------------------------------------------
INFO:omni: Completed createslice:

  Options as run:
                configfile: /home/chaos/tmp/portal5/omni_config
                framework: portal
                project: chaostest

  Args: createslice urn:publicid:IDN+ch5.gpolab.bbn.com:chaostest+slice+ecgtest2

  Result Summary: Created slice with Name urn:publicid:IDN+ch5.gpolab.bbn.com:chaostest+slice+ecgtest2, URN urn:publicid:IDN+ch5.gpolab.bbn.com:chaostest+slice+ecgtest2, Expiration 2013-05-16 04:53:55
 
INFO:omni: ============================================================
jericho,[~/tmp/portal5],16:53(0)$ 
```


_Trac comment by chaos on 05-15-2013 at 16:54_

**Comment by seledwards** [comment]
This means that everyone with an existing `omni_config` will have to edit or regenerate their `omni_config` using an updated version of `omni-configure.py`

_Trac comment by sedwards (github user: seledwards) on 05-15-2013 at 17:12_

**Comment by tcmitchell** [comment]
I see several issues here:
 1. User URN is incorrect in SliverStatus. This is the case on panther as well. The problem lies in the generated omni_config file.
  * Fix user URN in the generated file
  * Either remove the fake authority value or change framework from gcf to pgch -- the framework does not matter, it never gets used by omni as invoked by the portal
 1. The comment field in the generated ssh key contains the wrong hostname
  * Fix the comment field to simply contain the username, no hostname information
 1. `omni` deduces the wrong authority name from the URL of the SA. This involves the change from a hostname to a FQDN, which is desirable
  * To fix this, I think it will require an omni change.

Additionally, Chaos and I uncovered another issue:
 1. The service registry URL is constructed from the http host. It should instead be set in a config file. See `lib/php/sr_constants.php`

_Trac comment by tmitchel (github user: tcmitchell) on 05-15-2013 at 17:49_

**Comment by tcmitchell** [comment]
Updated status:
 1. The service registry issue is fixed on the tkt254-hostname branch.
 1. The SSH comment field has been changed to only include the username.
 1. The omni issues are being fixed for version 2.3

Remaining open issues:
 1. User URN is incorrect in SliverStatus. This is the case on panther as well. The problem lies in the generated omni_config file.
  - Fix user URN in the generated file
  - Fix the fake authority value in the generated file
  - Note: we cannot change to the pgch framework because the portal uses a version of gcf that predates the pgch framework


_Trac comment by tmitchel (github user: tcmitchell) on 05-22-2013 at 13:43_

**Comment by tcmitchell** [comment]
Since the incorrect URN (the lone remaining issue) is an existing issue on panther and all development hosts, we should detangle it from this ticket. It is identified as [proto-ch ticket 10](http://trac.gpolab.bbn.com/proto-ch/ticket/10).

_Trac comment by tmitchel (github user: tcmitchell) on 05-22-2013 at 14:38_

**Comment by ahelsing** [comment]
Tom: agreed, makes sense to track the incorrect URN issue on \[https://github.com/GENI-NSF/geni-portal/issues/10\].

_Trac comment by chaos on 05-22-2013 at 14:48_

**Comment by ahelsing** [comment]
I found another one, sorry: when Sarah asked to join my project, the e-mail i got was:
```
From: Sarah Edwards <sedwards@geni.net>
Date: Thu, 23 May 2013 15:31:14 -0400 (EDT)
To: Chaos Golubitsky<chaos@bbn.com>
Reply-To: sedwards@geni.net
Subject: Join GENI project chaostest?

May I join GENI project 'chaostest'?

I think I need to do GENI research in your project.

I am a student in your lab.

To handle my request, go to the GENI Portal here:
https://tau-ceti.gpolab.bbn.com/secure/handle-project-request.php?project_id=5f9e2637-9e3d-43df-920b-fc4dea7905dc&member_id=19ab582c-d4e9-4f63-9f07-d4ada46cdba4&request_id=1
```
That URL says tau-ceti.  It shouldn't.


_Trac comment by chaos on 05-23-2013 at 15:48_

**Comment by ahelsing** [comment]
I misidentified this bug (at least somewhat) --- Sarah turns out to have been browsing to `https://tau-ceti.gpolab.bbn.com/`.  That should either:
 A. Work as well as if she had been browsing to portal5.gpolab.bbn.com
 B. Not work at all

I vote for B.  i think we should just give every portal/CH a third IP address, so we can make sure that the hostname is not the same virtualhost as the base hostname.  Alternative is to make sure that `$_SERVER['SERVER_NAME']` is never used anywhere.  That might be desirable, but it's too hard for right now IMO.


_Trac comment by chaos on 05-23-2013 at 16:07_

**Comment by ahelsing** [comment]
Summarizing on this ticket, because i started on this ticket: for the time being, a reasonable fix is to use `NameVirtualHost` on our apache SSL sites, setup a `/var/www/localhost` containing a generic "this is not a good place for you to be" webpage, and make that the default site for both the portal-ssl and ch-ssl apache configurations.  I'm working on those changes now, and i think that will take care of this (unless it was masking other issues, but i don't _think_ my testing was susceptible to other issues like that, because i didn't have my browser configured to pass me through to https://tau-ceti.  Famous last words and all.)

_Trac comment by chaos on 05-23-2013 at 17:18_

**Comment by ahelsing** [comment]
I wanted to follow up and clarify something here, because i don't think i said this in so many words yesterday: in my opinion, the observed behavior, in which `$_SERVER['SERVER_NAME']` is used in the join request e-mail and other places which refer to portal web pages, is _not_ a portal bug.  The portal pages are successfully being served to clients who requested `SERVER_NAME`, so it's entirely reasonable for the portal pages to use that value as a name for itself.

(Where we've needed to remove `SERVER_NAME` references in the context of this ticket, it's because the references have been to clearinghouse services --- it's not okay for the portal to assume that CH services will share its own preferred name.)

As it turns out, we don't really want the portal webpages answering to more than one `SERVER_NAME`, because it's confusing and because it might leave us vulnerable to other (unknown) bugs which should be unimportant.  But the right place to fix that is in the webserver configuration.



_Trac comment by chaos on 05-24-2013 at 12:36_

**Comment by tcmitchell** [comment]
This task is complete and merged into develop for the 1.5 release.

_Trac comment by tmitchel (github user: tcmitchell) on 05-28-2013 at 14:47_
